sdcm/scan_operation_thread.py

```diff
@@ -1,163 +1,300 @@
+    AGG_COUNT_ALL = FullScanCommand("AGG_COUNT_ALL", Template("SELECT count(*) FROM $ks_cf$bypass_cache$timeout"))
+
+
+class FullscanException(Exception):
+    """ Exception during running a fullscan"""
+    ...
+
+
 # pylint: disable=too-many-instance-attributes
 class ScanOperationThread:
-    bypass_cache = ' bypass cache'
-    basic_query = 'select * from {}'
-
-    # pylint: disable=too-many-arguments,unused-argument
-    def __init__(self, db_cluster: [BaseScyllaCluster, BaseCluster], duration: int, interval: int,
-                 termination_event: threading.Event, ks_cf: str,
-                 scan_event: [FullScanEvent, FullPartitionScanReversedOrderEvent], page_size: int = 10000, **kwargs):
-        self.ks_cf = ks_cf
-        self.db_cluster = db_cluster
-        self.page_size = page_size
-        self.duration = duration
-        self.interval = interval
-        self.number_of_rows_read = 0
-        self.db_node = None
-        self.read_pages = 0
-        self.scans_counter = 0
-        self.time_elapsed = 0
-        self.total_scan_time = 0
-        self.scan_event = scan_event
-        self.termination_event = termination_event
+    """
+    Runs fullscan operations according to the parameters specified in the test
+    config yaml files. Has 4 main modes:
+    - random: uses a seeded random generator to generate a queue of fullscan
+    operations using all the available operation types
+    - table: uses only FullScanOperation
+    - partition: uses only FullPartitionScanOperation
+    - aggregate: uses only FullScanAggregatesOperation
+
+    Check FullScanParams class for parameters to tweak.
+
+    Note: the seeded random generator is shared between the thread class and
+    the operations instances. So if the operations queue is  Op1, Op2, Op3],
+    all of them will be initialized with the same random generator instance
+    as this thread. This way the CQL commands generated by the instances are
+    also replicable.
+    """
+
+    def __init__(self, fullscan_params: FullScanParams, thread_name: str = ""):
+        self.fullscan_params = fullscan_params
+        self.fullscan_stats = FullScanStats()
         self.log = logging.getLogger(self.__class__.__name__)
-        self.user = kwargs.get("user", None)
-        self.password = kwargs.get("password", None)
-        self._thread = threading.Thread(daemon=True, name=self.__class__.__name__, target=self.run)
-
-    def wait_until_user_table_exists(self, db_node, table_name: str = 'random', timeout_min: int = 20):
-        text = f'Waiting until {table_name} user table exists'
-        if table_name.lower() == 'random':
-            wait.wait_for(func=lambda: len(self.db_cluster.get_non_system_ks_cf_list(db_node)) > 0, step=60,
-                          text=text, timeout=60 * timeout_min, throw_exc=True)
+        nemesis_seed = self.fullscan_params.db_cluster.params.get("nemesis_seed")
```

Comment:
there's possibility that nemesis_seed is not set (not always using SisyphusMonkey), specify some random default.

Reply:
yes, I meant that but I was not precise.